File: app/src/main/java/org/gdgcd/demo/BooksAdapter.java
package org.gdgcd.demo;

import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.ViewGroup;

import org.gdgcd.demo.domain.Book;

import com.google.common.annotations.VisibleForTesting;

import java.util.ArrayList;
import java.util.List;

public class BooksAdapter extends RecyclerView.Adapter<BookViewHolder> {
    public BooksAdapter() {
        books = new ArrayList<>();
    }

    private List<Book> books;

    @Override
    public BookViewHolder onCreateViewHolder(ViewGroup viewGroup, int position) {
        return new BookViewHolder(LayoutInflater.from(viewGroup.getContext()).inflate(R.layout.book_card, viewGroup, false));
    }

    @Override
    public void onBindViewHolder(BookViewHolder viewHolder, int position) {
        viewHolder.updateView(books.get(position));
    }

    @Override
    public int getItemCount() {
        return books.size();
    }

    public void append(Book book) {
        int position = this.books.size();
        this.books.add(book);
        notifyInsert(position);
    }

    @VisibleForTesting
    public void notifyInsert(int position) {
        notifyItemInserted(position);
    }

}

File: app/src/main/kotlin/BooksAdapter.kt
package org.gdgcd.demo

import com.squareup.picasso
import com.squareup.picasso.Picasso
import org.gdgcd.demo
import org.gdgcd.demo.domain.Book
import java.util.ArrayList

public class BooksAdapter : android.support.v7.widget.RecyclerView.Adapter<BookItemViewHolder>() {

    private val books: MutableList<Book> = ArrayList()

    override fun onCreateViewHolder(viewGroup: android.view.ViewGroup, i: Int): BookItemViewHolder {
        return BookItemViewHolder(android.view.LayoutInflater.from(viewGroup.getContext()).inflate(demo.R.layout.book_card, viewGroup, false))
    }

    override fun onBindViewHolder(viewHolder: BookItemViewHolder, i: Int) {
        val book = books.get(i)
        Picasso.with(viewHolder.context)
                .load(book.imageUrl)
                .tag(viewHolder.context).into(viewHolder.coverImage)
        viewHolder.title.setText(book.title)

        viewHolder.ratingSub.unsubscribe()
    }

    override fun getItemCount(): Int {
        return books.size()
    }

    public fun append(book: Book) {
        val position = this.books.size()
        this.books.add(book)
        notifyItemInserted(position)
    }

}